portacl.5
.\" Copyright (c) 2023
.\"	Thomas Hurst
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
.Dd August 21, 2023
.Dt PORTACL 5
.Os
.Sh NAME
.Nm portacl
.Nd
.Xr rc 8
script for
.Xr mac_portacl 4
.Sh DESCRIPTION
.Nm
is an
.Xr rc 8
initialization script for
.Xr mac_portacl 4 ,
allowing administrators to limit or permit binding to local UDP and TCP ports
that would otherwise be privileged to the superuser.
.Pp
Configuration is via the following
.Xr rc.conf 5
variables.
.Bl -tag -width indent-two
.It Va portacl_enable
.Pq Vt bool
Set to
.Dq Li YES
to load and enable the
.Xr mac_portacl 4
module upon system initialization.
.It Va portacl_ Ns Ao Ar users|groups Ac
.Pq Vt str
A space-delimited list of user and group names or numeric IDs to apply
.Xr mac_portacl 4
rules for, using the associated
.Va portacl_ Ns Ao Ar user|group Ac Ns _ Ns Ao Ar ident Ac Ns _ Ns Ao Ar tcp|udp Ac
variables.
.It Va portacl_ Ns Ao Ar user|group Ac Ns _ Ns Ao Ar ident Ac Ns _ Ns Ao Ar tcp|udp Ac
.Pq Vt str
A space-delimited list of service names or port numbers to add to the
.Xr mac_portacl 4
ruleset for that user or group and protocol.
.It Va portacl_additional_rules
.Pq Vt str
A comma-delimited list of raw
.Xr mac_portacl 4
rules to include alongside those specified above.
.It Va portacl_port_high
.Pq Vt int
The highest port that should be considered privileged by
.Xr mac_portacl 4 .
This corresponds to the
.Xr sysctl 8
variable
.Va security.mac.portacl.port_high ,
and defaults to
.Pa 1023 .
Note that default privileged port restrictions will not be enforced above
this value.
.It Va portacl_suser_exempt
.Pq Vt bool
Allow the superuser to bind to all
.Xr mac_portacl 4
protected ports.
This corresponds to the
.Xr sysctl 8
variable
.Va security.mac.portacl.suser_exempt ,
and defaults to
.Pa YES .
.It Va portacl_autoport_exempt
.Pq Vt bool
Allow applications to use automatic binding to port 0.
This corresponds to the
.Xr sysctl 8
variable
.Va security.mac.portacl.autoport_exempt ,
and defaults to
.Pa YES .
.El
.Sh FILES
.Bl -tag -width ".Pa /etc/rc.conf" -compact
.It Pa /etc/rc.conf
.It Pa /etc/rc.conf.local
.El
.Sh EXAMPLES
Use
.Xr service 8
and
.Xr sysrc 8
to permit applications running under the
.Dq www
user to bind TCP ports http (80) and https (443).
.Bd -literal -offset indent
# sysrc portacl_users="www"
# sysrc portacl_user_www_tcp="http https"
# service portacl enable
# service portacl start
.Ed
.Sh SEE ALSO
.Xr mac_portacl 4 ,
.Xr rc.conf 5 ,
.Xr service 8 ,
.Xr sysrc 8
.Sh AUTHORS
.An Thomas Hurst .


portacl.sh
#!/bin/sh
#
# $FreeBSD$

# PROVIDE: portacl
# REQUIRE: FILESYSTEMS
# BEFORE: SERVERS
# KEYWORD: nojail

. /etc/rc.subr

name="portacl"
desc="Network port access control policy"
rcvar="portacl_enable"
extra_commands="reload printrules"
start_precmd="portacl_check_sysctl_conf"
start_cmd="portacl_start"
restart_cmd="portacl_start"
reload_cmd="portacl_start"
stop_cmd="portacl_stop"
printrules_cmd="portacl_printrules"
required_modules="mac_portacl"

: "${portacl_enable:="NO"}"
: "${portacl_port_high:="1023"}"
: "${portacl_suser_exempt:="YES"}"
: "${portacl_autoport_exempt:="YES"}"
: "${portacl_users:=""}"
: "${portacl_groups:=""}"
: "${portacl_additional_rules:=""}"

is_integer()
{
	case "${1}" in
	''|*[!0-9]*)
		return 1
		;;
	esac
}

# Split the argument on commas
split_comma()
{
	local rule
	local IFS=','
	for rule in $1
	do
		echo "${rule}"
	done
}

# Lookup port in /etc/services if it is not numeric
resolve_port()
{
	local lookup
	local port="$1"
	local proto="$2"

	if is_integer "${port}"; then
		echo "${port}"
		return 0
	fi

	case "${port}" in
	''|*[!a-z0-9_-]*)
		warn "invalid service name: ${port}"
		return 1
		;;
	esac

	lookup=$(awk -v service="${port}" -v proto="${proto}" '
		BEGIN {
			FS = "[ \t]+|/"
			found = 0
		}
		{
			sub(/#.*/, "", $0)
		}
		$1 == service && $3 == proto { found = 1 }
		$3 == proto && NF > 3 {
			for (i = 4; i <= NF; i++) {
				if ($i == service) {
					found = 1
					break
				}
			}
		}
		found == 1 {
			print $2
			exit
		}' /etc/services)

	if [ -z "${lookup}" ]; then
		warn "unknown service ${port}"
		return 1
	fi

	echo "${lookup}"
}

list_rules_for()
{
	local id ident ident_list port port_list proto
	local kind="$1"
	local key="uid"
	local idflag="-u"

	if [ "${kind}" = "group" ]; then
		key="gid"
		idflag="-g"
	fi

	eval ident_list="\${${name}_${kind}s}"
	for ident in ${ident_list}
	do
		id=$(${ID} "${idflag}" "${ident}" 2>/dev/null)
		if [ -z "${id}" ]; then
			warn "unknown ${kind} ${ident}"
			continue
		fi

		for proto in tcp udp
		do
			eval port_list="\${${name}_${kind}_${ident}_${proto}}"
			for port in ${port_list}
			do
				port=$(resolve_port "${port}" "${proto}")
				[ -z "${port}" ] && continue
				echo "${key}:${id}:${proto}:${port}"
			done
		done
	done
}

# list all rules, one per line
list_rules()
{
	list_rules_for user
	list_rules_for group
	[ -n "${portacl_additional_rules}" ] && split_comma "${portacl_additional_rules}"
}

# generate a complete, validated ruleset ready for mac_portacl
generate_rules()
{
	list_rules | sort -ut : | tr '\n' , | validate_rules
}

# Filter out invalid rules and warn on stderr
validate_rules()
{
	awk '
		BEGIN { RS=","; FS=":"; sep = ""; len=-1 }
		{
			newlen=len + length($0) + 1
			if (newlen > 1023) {
				print "WARNING: reached portacl rule limit of 1023 bytes" > "/dev/stderr"
				exit 1
			}
			if (NF == 4 &&
			    ($1 ~ /^(uid|gid)$/) &&
			    ($2 ~ /^[0-9]+$/ && $2 >= 0 && $2 <= 65535) &&
			    ($3 ~ /^(tcp|udp)$/) &&
			    ($4 ~ /^[0-9]+$/ && $4 >= 0 && $4 <= 65535)) 
			{
				len=newlen
				printf("%s%s", sep, $0)
				sep=","
			} else {
				print "WARNING: Invalid portacl rule:", $0 > "/dev/stderr"
			}
		}
	'
}

portacl_printrules()
{
	generate_rules
	echo
}

portacl_check_sysctl_conf()
{
	local f overrides oid

	for f in /etc/sysctl.conf /etc/sysctl.conf.local
	do
		[ -r "${f}" ] || continue

		overrides="$(awk -F= '
			BEGIN { ORS=" " } 
			$1 ~ /^(security\.mac\.portacl\.|net\.inet\.ip\.portrange\.reserved)/ {
				print $1
			}
		' "${f}")"

		for oid in $overrides
		do
			warn "overriding ${oid} in ${f}"
		done
	done
}

portacl_start()
{
	local rules="$(generate_rules)"
	local port_high=1023
	local suser_exempt=1
	local autoport_exempt=1

	if is_integer "${portacl_port_high}"; then
		port_high="${portacl_port_high}"
	else
		warn "\$portacl_port_high is not set properly - see rc.conf(5)"
	fi

	checkyesno portacl_suser_exempt || suser_exempt=0
	checkyesno portacl_autoport_exempt || autoport_exempt=0

	${SYSCTL} security.mac.portacl.rules="${rules}" >/dev/null &&
	${SYSCTL} security.mac.portacl.suser_exempt="${suser_exempt}" >/dev/null &&
	${SYSCTL} security.mac.portacl.autoport_exempt="${autoport_exempt}" >/dev/null &&
	${SYSCTL} security.mac.portacl.port_high="${port_high}" >/dev/null &&
	${SYSCTL} security.mac.portacl.enabled=1 >/dev/null &&
	${SYSCTL} net.inet.ip.portrange.reservedlow=0 >/dev/null &&
	${SYSCTL} net.inet.ip.portrange.reservedhigh=0 >/dev/null
}

portacl_stop()
{
	${SYSCTL} net.inet.ip.portrange.reservedlow=0 >/dev/null &&
	${SYSCTL} net.inet.ip.portrange.reservedhigh=1023 >/dev/null &&
	${SYSCTL} security.mac.portacl.enabled=0 >/dev/null
}

load_rc_config $name
run_rc_command "$1"
